components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -1812,125 +1816,318 @@ private int handleStartIndexEqualsNULL(Integer startIndex) {
 
+        String attributeName = node.getAttributeValue();
+        String filterOperation = node.getOperation();
+        String attributeValue = node.getValue();
         if (log.isDebugEnabled()) {
-            log.debug("Listing groups with filter: " + attributeName + filterOperation +
-                    attributeValue);
-        }
+            log.debug("Filtering groups with filter: " + attributeName + " + " + filterOperation + " + "
+                    + attributeValue);
+        }
+        // Check whether the filter operation is supported for filtering in groups.
+        if (isFilteringNotSupported(filterOperation)){
+            String errorMessage = "Filter operation: " + filterOperation + " is not supported for groups filtering.";
+            throw new CharonException(errorMessage);
+        }
+        // Resolve the domain name in request according to 'FilterUsersAndGroupsOnlyFromPrimaryDomain' or
+        // EnableFilteringEnhancements' properties in identity.xml or domain name embedded in the filter attribute
+        // value.
+        domainName = resolveDomain(domainName, node);
         List<Object> filteredGroups = new ArrayList<>();
-        //0th index is to store total number of results;
+        // 0th index is to store total number of results.
         filteredGroups.add(0);
         try {
-            String[] roleList = getGroupList(attributeName, filterOperation, attributeValue);
+            String[] roleList = getGroupList(node, domainName);
             if (roleList != null) {
                 for (String roleName : roleList) {
                     if (roleName != null && carbonUM.isExistingRole(roleName, false)) {
-                        //skip internal roles
-                        if ((CarbonConstants.REGISTRY_ANONNYMOUS_ROLE_NAME.equals(roleName)) ||
-                                UserCoreUtil.isEveryoneRole(roleName, carbonUM.getRealmConfiguration())) {
+                        // Skip internal roles.
+                        if (CarbonConstants.REGISTRY_ANONNYMOUS_ROLE_NAME.equals(roleName) || UserCoreUtil
+                                .isEveryoneRole(roleName, carbonUM.getRealmConfiguration())) {
                             continue;
                         }
-                        String userStoreDomainName = IdentityUtil.extractDomainFromName(roleName);
-                        if (isInternalOrApplicationGroup(userStoreDomainName) || isSCIMEnabled(userStoreDomainName)) {
-                            if (log.isDebugEnabled()) {
-                                log.debug("SCIM is enabled for the user-store domain : " + userStoreDomainName + ". " +
-                                        "Including group with name : " + roleName + " in the response.");
-                            }
-                            Group group = getGroupWithName(roleName);
+                        Group group = getRoleWithDefaultAttributes(roleName);
+                        if (group != null) {
```

Comment:
We have to check if there is an Id associated to the group as well right

Reply:
Fixed: 9d79afc6848cebbb4126786fc6eb56cd8d2dd7fd